Add balance to api v1 customers endpoint

mickenorlen opened the pull request:
### Update

After first round of reviews: 
- PR description updated to reflect new solution.
- Commit history improved and cleaned to avoid unnecessary history rewrites.  

### Why?

- Closes #8909

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->

Add balance to api v1 customers endpoint
    
- `customers#show`: Add balance (type: double) to customer attributes.
- `customers#index`: Add balance only if specified in query parameter: `extra_fields[customer]=balance`

### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- Visit Swagger UI: http://localhost:3000/api-docs
- Try "List Customers" with and without `extra_fields[customer]=balance`
- Try "Show Customer"
- See that balance is added appropriately
- Check also the "Schema" For each request and global Schemas at end of page to see that they are set correctly

### Release notes

Changelog Category: User facing changes

mickenorlen commented:
I'm leaving 1 rubocop lint error as the file is defining long schema definitions. Let me know if this is not acceptable.
`Metrics/ClassLength: Class has too many lines. [130/100]`
https://github.com/openfoodfoundation/openfoodnetwork/pull/10330/files#diff-5203fba163c2f40172654685c940e5af5f5eff3243931474aeaad04c40c82847

mkllnk commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -10,15 +10,29 @@ class CustomersController < Api::V1::BaseController
         @pagy, customers = pagy(search_customers, pagy_options)
-
-        render json: Api::V1::CustomerSerializer.new(customers, pagination_options)
+        render json: Api::V1::CustomerSerializer.new(
+          customers,
+          pagination_options.merge({
+                                     params: {
+                                       display_customer_balance: @display_customer_balance
+                                     }
+                                   })
+        )
```
To answer your style question. I would format it like this:
```suggestion
        render json: Api::V1::CustomerSerializer.new(
          customers,
          pagination_options.merge(
            params: { display_customer_balance: @display_customer_balance }
          )
        )
```

Well, actually, in this case I would simplify it further:


```suggestion
        options = pagination_options.merge(
          params: { display_customer_balance: @display_customer_balance }
        )
        render json: Api::V1::CustomerSerializer.new(customers, options)
```

But, as said above, I would hope not to need this option at all.

mkllnk commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -54,13 +68,28 @@ def customer
       end
 
+      def set_display_customer_balance
+        val = params[:display_customer_balance]
+        return if val.nil?
+
+        unless val.in?(["true", "false", "1", "0"])
+          invalid_query_parameter(:display_customer_balance, :unprocessable_entity, "Not a boolean")
+        end
+        @display_customer_balance = ActiveModel::Type::Boolean.new.cast(val)
```
Most of this could be in a re-usable method. The text "Not a boolean" should be translated via `I18n`.

mickenorlen replied:
Fixed with `Api::V1::QueryParam` and related translations, now for `extra_fields` instead of `boolean`

mkllnk commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -54,13 +68,28 @@ def customer
         customers = visible_customers.includes(:bill_address, :ship_address)
         customers = customers.where(enterprise_id: params[:enterprise_id]) if params[:enterprise_id]
+
+        if @display_customer_balance
+          customers = CustomersWithBalance.new(customers: customers).query
+        end
```
This is generating a complete new query based on customer ids. It resets the `includes(:bill_address, :ship_address)` and could have a negative impact on performance. Wouldn't it be better to take the existing query and add the customer balance part on top?

`CustomersWithBalance` could be greatly simplified by taking a customers relation in the initializer. Other places can then call it with `Customer.of(enterprise)`. Much better division of responsibilities.

mkllnk commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -70,6 +99,10 @@ def visible_customers
         )
       end
 
+      def editable_enterprises
+        OpenFoodNetwork::Permissions.new(current_api_user).editable_enterprises.select(:id)
+      end
```
Why did you move this method?

mickenorlen replied:
It made more sense for the order of method calls and to have params at bottom, guess it wasnt worth it for rewriting history.

mkllnk commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -10,15 +10,29 @@ class CustomersController < Api::V1::BaseController
+        render json: Api::V1::CustomerSerializer.new(
+          CustomersWithBalance.new(customers: customer).query.first,
+          {
+            include: [params.fetch(:include, [])].flatten.map(&:to_s),
+            params: { display_customer_balance: true },
```
This API is following JSON:API. There is a standardised way to include attributes like this with the `fields` option. We should probably use that. Have a look at https://jsonapi.org/examples/.

Using the `fields` option hides all other fields but I think that's acceptable for our use case. The API call can mention all desired fields. Is that okay @lin-d-hop?

lin-d-hop replied:
Yes if that is standard practice then I say we go with that :+1:

mkllnk commented on app/json_schemas/customer_schema.rb:
```diff
@@ -61,4 +61,18 @@ def self.writable_attributes
+        required: [:value, :time],
+        properties: {
+          value: { type: :number, format: :double },
+          time: { type: :string, format: "date-time" }
```
I'm not sure about including the time within a new object. That's a cool idea and knowing the time could be really helpful but maybe it would be even more helpful to have a new field on the customer itself, for example `viewed_at`. Other properties can change over time as well. And if you see the whole thing like a "report" then it makes sense to have that attribute on the top level.

On the other hand, the querying program can just note the timestamp of retrieval as well.

What do you think, @lin-d-hop?

lin-d-hop replied:
I guess we'd find it helpful for debugging to include the time, but I agree @mkllnk that it can be on the top level. I don't have a strong opinion here.

lin-d-hop replied:
Responded in comment below.

mickenorlen replied:
Time removed

mkllnk commented on spec/requests/api/v1/customers_spec.rb:
```diff
@@ -105,6 +107,56 @@
+        end
+
+        it "returns unprocessable entity for non boolean values" do
+          get "/api/v1/customers", params: { display_customer_balance: "unknown" }
+          expect(response.status).to eq 422
```
We may need to check the error message here because to make sure we are catching the right error.

mickenorlen replied:
Fixed! Was surprised to see that method missing caused 422 instead of internal server error

mkllnk commented on swagger/v1/swagger.yaml:
```diff
@@ -77,7 +77,7 @@ components:
                   type: object
                   nullable: true
-                  example:
+                  example: 
```
These changes seem unrelated and don't make any sense. Are newer version of Swagger causing these? It's good to commit unrelated changes separately.

mickenorlen replied:
I dont know why this happened. I can split it if you want

mkllnk reviewed (changes requested):
Thank you for this contribution. This is a great step forward. I would have liked to read it in smaller commits though and I left a few comments about details below. But it's almost there! Well done.

lin-d-hop commented:
Hi @mickenorlen 
Thanks for this contribution!

> I added the balance time to make things more explicit. But I've also been thinking that maybe in the future, it would be nice to > be able to specify multiple times for which the balance is to be retrieved. This would remove the need of scheduling requests > at certain times and reduce the amount of requests.
>
> Eg: instead of setting up a schedule to request the balance every day, you could send a request once a month getting the 
> balance for each day of that month.

Regarding this proposal, this feels to me like over-engineering at this stage. The use case you describe is more something related to reports. Since the customer balance is a calculated field in OFN, and very resource intensive to calculate, I do not want to be making it easier for people to request a bunch of balances at once. Instead, let's treat the customer endpoint as a snapshot. The API is an automation tool so setting up scheduled tasks feels okay to me. If a use case emerges to give historical balances then we can solve that use case then, likely with a report rather than this endpoint.

mickenorlen commented:
@mkllnk @lin-d-hop

Thank you for your feedback. I feel like we should agree on the `fields` path before I start responding to individual comments. Below is mostly a technical discussion for devs. 

**TLDR:** I see no problems using fields for this type of problem. But we would probably have to implement this currently missing fields functionality across the board to be consistent and add some documentation on it.  Let me know if you think we should open a separate PR for adding fields before continuing with this one.

---

## Fields
After some reading I found a discussion on our use case at jsonapi:
> Costly lookups that we’d like to avoid unless the user is really asking for the information:

https://discuss.jsonapi.org/t/must-i-return-all-resource-object-attributes-without-a-sparse-fieldset-specified/486/9

And indeed they quite recently updated the specification to allow for the `fields` query to be used in this way. So although `sparse fields`, as the name implies, was usually intended for getting a filtered subset of fields. It can also be used to show otherwise hidden ones.

However, the comment linked above also mentions: 
>  It works nicely, but since it’s not in the spec, we had to build this atop of the jsonapi library we were using, and addressing all the recursive cases was a little involved.

## Implementation draft
Our `jsonapi-serializer` accepts the `fields` option as expected but does not care how these fields are derived. As we aren't passing any `fields` to the serializer, requesting fields is not yet supported at all (eg `?fields[customer]=balance`).

So in order to support it we would have to implement something like this:

```RUBY
# customers_controller
def index
  # ...
  if customer_fields.include(:balance)
    add_balance_to_customers
  end

  serializer = Api::V1::CustomerSerializer.new(customers, options.merge(customer_fields))
  # ...
end

def show
  # ...
  serializer = Api::V1::CustomerSerializer.new(customers, options.merge(customer_fields(customer_fields_for_show)))
  # ...
end

# Prefer always returning set of default fields instead of having to check if present
def customer_fields(fallback = Api::V1::CustomerSerializer.default_fields)
  # fields is comma separated list https://jsonapi.org/format/#fetching-sparse-fieldsets
  fields = params.dig(:fields, :customer)&.split(', ').presence || fallback
  {fields: {customer: fields}}
end

# In show we add that extra field
def customer_fields_for_show
  Api::V1::CustomerSerializer.default_fields + [:balance]
end
```
```RUBY
# customer_serializer
def self.default_fields
  attributes_to_serialize.keys.except(:balance)
end
```

mkllnk commented:
Thank you for doing the research and giving it this much thought. That's excellent work.


>     As the jsonapi specification was only recently updated to kind of vaguely allow for this type use case - It doesn't really feel like standard or expected behavior.

Good point. Let's say it's an accepted behaviour. Others may do the same but there is no well-established standard for thi s problem, right?

>     I'm not sure it integrates well with the swagger api documentation. How would you inform the user of these hidden attributes and that they have to use the fields option to find them?

Good point as well. We could add it to the examples. But I found that Swagger only documents one (the last) example in the specs. I haven't seen a way to document several examples for the same endpoint with the same result. Well, I didn't search very hard either. Do you know how a way?

>     Like you mentioned users would have to specify every other attribute as well. Which if made standard for the api could make it feel a bit clunky to work with. What is the benefit of fields?

The attributes would only need to be specified when the fields parameter is given. Otherwise we use standard fields.

Since we don't support the fields yet but they could be very useful, I was hoping that this could be a good opportunity to add that functionality. I'm a bit unsure now. So I think the two options are:

* Use a standardised feature in a "non-standard but accepted" way to display customer balance. This sounds more robust and future-proof to me but is less convenient for creating queries and for documentation.
* Use our own parameter (family) which would be easier to use but has the risk of not getting it right and discovering mistakes in our naming or architecture later.

In terms of implementation of customer balance. Since we only support the show method, we could calculate the balance in a separate query when the serializer calls that attribute. It doesn't have to be in the same query.

Okay, since I'm undecisive, let's ask another developer like @dacook for his opinion. And if you don't have a clear opinion either then the author of the pull request can choose. :smile: You've done some excellent work already and I think that you would find a good solution.

dacook commented:
Hi @mickenorlen , as Maikel already said, thank you for your work on this solution! I appreciate the options and discussion presented here. After looking into it and discussing a bit with Maikel, I can see that:

The jsonapi "[**sparse fieldsets**](https://jsonapi.org/format/#fetching-sparse-fieldsets)" (`fields`) specification means that it should filter to only output the listed attributes. 
According to that specification, you would need to list ALL desired attributes, like so: 
```
/api/v1/customers/?fields[customer]=id,enterprise_id,first_name,...,balance
```

I don't think we should implement our own behaviour, because we might one day want to use the sparse fieldsets behaviour, or it might conflict with functionality in a library we use (ie https://github.com/jsonapi-serializer/jsonapi-serializer#sparse-fieldsets).

So we agreed we should create the new behaviour that you've drafted, with a similar name, such as: 
```
/api/v1/customers/?extra_fields[customer]=balance
```

Can you please go ahead with that format?
Also, in your draft I see a comma-space separator `", "`. It would be better to mimic the sparse fieldsets spec of just comma only `","`.

mickenorlen commented:
Thank you for your positive comments :smile:  I like your solution with `extra_fields`, will start working on it. Should indeed be comma only. You both have good eyes to spot code errors!

Just wanted to check in with you on this idea. Not sure it came across.
> Could it be an idea to add a balance field to the customer table and update it automatically when orders change their statuses?

This could make `balance` a non-calculated field and remove the need for hiding this attribute to begin with.

mkllnk commented:
>  Could it be an idea to add a balance field to the customer table and update it automatically when orders change their statuses?
> 
> This could make balance a non-calculated field and remove the need for hiding this attribute to begin with.

Good point. But I would like to avoid that at the moment.

* > There are only two hard things in Computer Science: cache invalidation and naming things.
  > -- Phil Karlton
  
  https://martinfowler.com/bliki/TwoHardThings.html
* We inherited many callbacks from Spree which make it really hard for us to control when things should get updated or not. For example, updating shipping fees when a shipping method is updated or deleted turns out to be very complicated and entangled with other logic updating adjustments. If we can simplify our code there then we may be ready for caching.

mickenorlen commented on app/queries/customers_with_balance.rb:
```diff
@@ -19,24 +18,23 @@ def query
 
   def validate_arguments
-    return unless [enterprise, @customers].all?(&:blank?)
+    return unless [enterprise, @customers].all?(&:nil?)
```
Noticed that an empty customer array could throw this exception. Fixed and added test

mickenorlen commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -6,31 +6,28 @@ module Api
   module V1
     class CustomersController < Api::V1::BaseController
       include AddressTransformation
+      include Api::V1::QueryParam
```
Extracted query param validation logic to included module `Api::V1::QueryParam`
I made it as simple as I could + added spec

mkllnk replied:
Good one. :+1:

mickenorlen commented on app/serializers/api/v1/customer_serializer.rb:
```diff
@@ -16,11 +16,9 @@ class CustomerSerializer < BaseSerializer
-      } do |object|
-        { value: object.balance_value, time: object.balance_time }
-      end
+      attribute :balance, if: proc { |record|
```
Now we just check if customer responds to balance_value, no params passed as you suggested

mickenorlen commented on app/serializers/api/v1/customer_serializer.rb:
```diff
@@ -16,11 +16,9 @@ class CustomerSerializer < BaseSerializer
-        params && params[:display_customer_balance] == true
-      } do |object|
-        { value: object.balance_value, time: object.balance_time }
-      end
```
Now we just check if customer responds to balance_value, no params passed as you suggested

mickenorlen commented on app/serializers/api/v1/customer_serializer.rb:
```diff
@@ -16,11 +16,9 @@ class CustomerSerializer < BaseSerializer
-      end
+      attribute :balance, if: proc { |record|
+        record.respond_to?(:balance_value)
+      }, &:balance_value
```
Now we just check if customer responds to balance_value, no params passed as you suggested

mickenorlen commented on spec/models/api/v1/query_param_spec.rb:
```diff
@@ -0,0 +1,60 @@
+require 'spec_helper'
+
+describe Api::V1::QueryParam do
+  let(:dummy_controller) { Api::V1::BaseController.new.extend Api::V1::QueryParam }
```
I'm still reusing `json_api_error` from base_controller. Could be an idea to extract all error handeling to a separate module, but felt a bit too much for now

mickenorlen commented on app/models/api/v1/query_param.rb:
```diff
@@ -0,0 +1,34 @@
+  module V1
+    module QueryParam
+      def invalid_query_param(name, status, msg)
+        render status: status, json: json_api_error(msg, error_options:
```
I'm still reusing json_api_error from base_controller. Could be an idea to extract all error handeling to a separate module, but felt a bit too much for now

mickenorlen replied:
I'm still reusing `json_api_error` from base_controller. Could be an idea to extract all error handeling to a separate module, but felt a bit too much for now

mickenorlen commented on app/serializers/api/v1/customer_serializer.rb:
```diff
@@ -16,6 +16,10 @@ class CustomerSerializer < BaseSerializer
 
+      attribute :balance, if: proc { |record|
+        record.respond_to?(:balance_value)
+      }, &:balance_value
```
Now we just check if customer responds to `balance_value`, no params passed as you suggested

mkllnk replied:
Excellent!

mickenorlen commented on app/controllers/api/v1/base_controller.rb:
```diff
@@ -114,6 +114,10 @@ def json_api_invalid(message, errors)
+
+      def validate_query_parameter(name, status, msg)
+        render json_api_error(msg, status: status, source: { parameter: name })
+      end
```
Fixed and refactored to `Api::V1::QueryParam`

mickenorlen commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -10,15 +10,29 @@ class CustomersController < Api::V1::BaseController
 
       before_action :authorize_action, only: [:show, :update, :destroy]
+      # Query parameters
+      before_action :set_display_customer_balance, only: [:index] # Always show balance for show
```
This style is to avoid double render error which is not thrown if validation is performed in before_action

mickenorlen commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -10,15 +10,29 @@ class CustomersController < Api::V1::BaseController
+                                       display_customer_balance: @display_customer_balance
+                                     }
+                                   })
+        )
```
Good idea, implemented

mickenorlen commented:
@mkllnk I got a bit confused after our last discussion about how to add commits into a PR in review.

Now I just added a big commit with the new changes for this round and added some comments in the code to help you out a bit. It got a bit late and I'm not sure what type of history we want to merge in the end.

Let me know if you want me to sit down tomorrow and try to separate it a bit more.

mkllnk commented on spec/models/api/v1/query_param_spec.rb:
```diff
@@ -0,0 +1,60 @@
+require 'spec_helper'
+
+describe Api::V1::QueryParam do
+  let(:dummy_controller) { Api::V1::BaseController.new.extend Api::V1::QueryParam }
```
Why extend instead of include?

mickenorlen replied:
The result is the same. But updated with include syntax as agree is clearer

mkllnk commented on app/models/api/v1/query_param.rb:
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+# To be included in api controllers for handeling query params
```
A module to be included in a controller is usually called a *concern*. Have a look at this one:
https://github.com/openfoodfoundation/openfoodnetwork/blob/e066e5ca74d8e6ffe626972cba2d4b32377df587/app/controllers/concerns/raising_parameters.rb#L1-L23

mickenorlen replied:
Fixed

mkllnk commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -6,19 +6,30 @@ module Api
+        render json: Api::V1::CustomerSerializer.new(
+          CustomersWithBalance.new(customers: customer).query.first,
+          {
+            include: [params.fetch(:include, [])].flatten.map(&:to_s),
```
I found the `include_options` easier to read.

mkllnk commented on app/json_schemas/customer_schema.rb:
```diff
@@ -61,4 +61,9 @@ def self.writable_attributes
     [:enterprise]
   end
+
+  # Optional attributes included with "with"
```
Do we need to adjust the comment here?

mickenorlen replied:
Fixed

mkllnk commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -61,6 +72,11 @@ def authorize_action
         customers = visible_customers.includes(:bill_address, :ship_address)
         customers = customers.where(enterprise_id: params[:enterprise_id]) if params[:enterprise_id]
+
+        if @extra_customer_fields.include?(:balance)
+          customers = CustomersWithBalance.new(customers: customers).query
+        end
```
Repeating a comment from my first review: This is generating a complete new query based on customer ids. It resets the `includes(:bill_address, :ship_address)` and could have a negative impact on performance. Wouldn't it be better to take the existing query and add the customer balance part on top?

mickenorlen replied:
Missed that one. Very good thinking, fixed, much cleaner now

mkllnk commented on app/queries/customers_with_balance.rb:
```diff
@@ -1,26 +1,41 @@
+# Adds an aggregated 'balance_value' to each customer based on their order history
+#
 class CustomersWithBalance
-  def initialize(enterprise)
+  def initialize(
+    enterprise: nil,  # Filter customers by enterprise
+    customers: nil    # Filter customers by record/collection of customers/ids
+  )
     @enterprise = enterprise
+    @customers = customers
+
+    validate_arguments
```
We can simplify this by just taking a customers relation in the initializer. This module wouldn't need to know `enterprise` at all. Other places can then call it with `Customer.of(enterprise)`. Much better division of responsibilities.

mickenorlen replied:
Fixed

mkllnk commented on spec/models/api/v1/query_param_spec.rb:
```diff
@@ -0,0 +1,60 @@
+              detail: "error message",
+              source: { parameter: "param" },
+              status: :unprocessable_entity,
+              title: I18n.t("api.query_param.error.title")
```
Better to spell it out in plain text. https://github.com/openfoodfoundation/openfoodnetwork/wiki/Code-Conventions#prefer-plain-text-over-method-calls-in-expected-values

mickenorlen replied:
Fixed

mkllnk reviewed (changes requested):
Good work.

> I got a bit confused after our last discussion about how to add commits into a PR in review.

My favourite way is to add one commit for each review comment. As a reviewer, I can then look at each commit and mark the related comment as resolved if I'm happy with the solution.

Since I reviewed it now, you don't need to split it all again for me. But there are some detours in the commits now and we will need a second review by another developer once I approve this. So maybe try to do smaller commits going forward and then once I approve, you can also split the big commits, remove detours and the next reviewer will have a much easier task. Code and even the git history is read many times. It's worth putting some effort in.

mickenorlen commented on app/models/api/v1/query_param.rb:
```diff
@@ -0,0 +1,34 @@
+# To be included in api controllers for handeling query params
+module Api
+  module V1
+    module QueryParam
```
Naming feels difficult. Converted to concern named `ReusableQueryParams`.

mickenorlen commented:
> We inherited many callbacks from Spree which make it really hard for us to control when things should get updated or not. For example, updating shipping fees when a shipping method is updated or deleted turns out to be very complicated and entangled with other logic updating adjustments. If we can simplify our code there then we may be ready for caching. At the moment it would be really hard to update a balance when the amount of an adjustment is updated in a transition of the state machine. And what if an error is raised?

I understand there is a lot of complexity behind orders. But I'm not sure I understand why it would be relevant for this problem as we are simply checking values in the orders table. 

Given the current balance query
```RUBY
class OutstandingBalance
  def statement
    <<-SQL.strip_heredoc
      CASE WHEN "spree_orders"."state" IN #{non_fulfilled_states_group.to_sql} THEN "spree_orders"."payment_total"
           WHEN "spree_orders"."state" IS NOT NULL THEN "spree_orders"."payment_total" - "spree_orders"."total"
      ELSE 0 END
    SQL
  end
end
```

Wouldn't we have access to the exact same values by hooking in to `after_update` in rails spree_order?

```RUBY
module Spree
  class Order < ApplicationRecord
  
    after_update :update_customer_balance
    
    def update_customer_balance
      if saved_change_to_attribute?(:state) || 
        saved_change_to_attribute?(:payment_total) || 
        saved_change_to_attribute?(:total)
        
        # these would be similar to outstanding balance calculation
        previous_order_balance_value = get_previous_order_balance_value 
        new_order_balance_value = get_new_order_balance_value
        diff = new_order_balance_value - previous_order_balance_value
        
        if diff != 0
          customer.update(:balance, customer.balance + diff)
        end
      end
    end
  
  end
end
```

This could be added on trial to run in parallell with current implementation until it's been verified to update balance corectly:
1. create balance column, add current customer balances and balance update logic
2. come back later to see that customer.balance has the correct value
3. start actually using column

mkllnk commented:
> we are simply checking values in the orders table.
> Wouldn't we have access to the exact same values by hooking in to after_update in rails spree_order?

Your solution looks like it could work but it could also go wrong from time to time. For example a stale record could mean that the update fails and then the value is out of sync. You can solve that problem but how much complexity do we want to add to solve this?

I agree that it would be nice to have the customer balance stored at some point. I just think that we need to work on a lot more tech debt before introducing another cached value. I see that we are adding complexity to the API instead. That's not great either but it's isolated. The API code doesn't affect any other business logic. I'm trying to reduce the potential for bugs.

mkllnk commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -6,7 +6,7 @@ module Api
     class CustomersController < Api::V1::BaseController
       include AddressTransformation
-      include Api::V1::QueryParam
+      include ReusableQueryParams
```
What about `ExtraFields`?

mickenorlen replied:
Sure

mickenorlen replied:
Sure, I guess I thought it could become a small library. But I'm starting to understand your logic of "its not a library yet"

mkllnk commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -25,7 +25,7 @@ def index
       def show
         render json: Api::V1::CustomerSerializer.new(
-          CustomersWithBalance.new(customers: customer).query.first,
+          CustomersWithBalance.new(Customer.where(id: params[:id])).query.first,
```
Hm, so we already have a customer record and reload the whole think just to add the balance? I'm wondering if it would be easier to have a method on the the customer model. But let's leave that for now because that would have implications on the serializer again because the attribute would always be available.

mkllnk commented on spec/controllers/concerns/reusable_query_params_spec.rb:
```diff
@@ -3,7 +3,7 @@
 
 describe ReusableQueryParams do
-  let(:dummy_controller) { Api::V1::BaseController.new.extend ReusableQueryParams }
+  let(:dummy_controller) { Api::V1::BaseController.include(ReusableQueryParams).new }
```
Now I see. For `include` you need to change the BaseController. That's a side-effect which leaks into other specs, right? Then it's better to just extend the instance instead of modifying the class. I'm sorry that I didn't realise that.

mickenorlen replied:
Yes, it leaks, didn't think about that. I'm sure there is some syntax with using include if include hooks are used. But as of now extend works.

lin-d-hop commented:
> I agree that it would be nice to have the customer balance stored at some point. I just think that we need to work on a lot more tech debt before introducing another cached value. I see that we are adding complexity to the API instead. That's not great either but it's isolated. The API code doesn't affect any other business logic. I'm trying to reduce the potential for bugs.

Thanks @mkllnk I agree. We can't introduce another cached value for balance right now. Despite how annoying it is in resource heaviness, this value is too critical to introduce scope for error. Having a correct value for customer balance in the API is critical as this will be used by users to generate invoices, balance due notices and accounting/VAT reports for enterprises. Errors in this figure will create huge distrust in the system and this is something we can afford less than high resource consumption and isolated code complexity.

@mickenorlen It's great that you are generating creative responses! But let's focus on getting this wrapped up in line with the original spec :)

mkllnk reviewed (changes requested):
Cool, this is so much better. Thank you for creating small commits. That made my review much easier. There's only one thing I'd like to see reverted again. Sorry!

mickenorlen commented:
I know I'm asking a lot of questions. Thank you for your patience! Once I learn more about the application and your way of working, I hope to become more seamless.

@mkllnk As you mentioned, I'm thinking to rewrite a clean commit history for the next reviewer after you approve.

mkllnk commented on spec/controllers/concerns/extra_fields_spec.rb:
```diff
@@ -2,8 +2,8 @@
-describe ReusableQueryParams do
-  let(:dummy_controller) { Api::V1::BaseController.include(ReusableQueryParams).new }
+describe ExtraFields do
+  let(:dummy_controller) { Api::V1::BaseController.new { include ExtraFields }}
```
Perfect!

mkllnk reviewed (approved):
Great. Ping me when you are done rewriting the history and I'll move this back to the Code Review column for others to review.

mickenorlen commented:
@mkllnk 

- [x] History cleaned

While cleaning I found and fixed a couple of minor inconsistencies. Summary below, diff before/after [here](https://github.com/mickenorlen/openfoodnetwork/compare/a1bd60a3751651e4c555809e89e310df99e9b9cc..b749ca8ffcf68b6368c5ca258b0f85876523f421).
- `JsonApiSchema`. Rename option `with` to `extra_fields` for consistency with `extra_fields` name of new query param. 
- `customers_with_balance_spec.rb`: Undo outdated variable name change and rm no longer needed method to reduce git changes to [essential diff](https://github.com/openfoodfoundation/openfoodnetwork/pull/10330/files#diff-425147f8d02a000f10338f539afcc964560a6bf9b607d2481796d80732b161f2) only. 

Also, this [solution](https://github.com/openfoodfoundation/openfoodnetwork/pull/10330#discussion_r1102124373)  you commented on was not supposed to have been committed as it doesn't work. Sorry about that. The final change was a full revert to first version using extend, [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/10330/files#diff-ace2d9f00f64cf6c051dbf7ce2e409593aabda2fc9fff0fc188df513d277ba95R6).

mickenorlen commented:
Unrelated tests failing even after rebase to upstream/master

mkllnk reviewed (commented):
Great, thank you. I noticed one Rubocop issue and I'm wondering what the best solution here is. We should probably split the file because it's becoming to big. One good separation could be to have the data like hashes in a separate class and keep the logic in this one.

mickenorlen commented:
Sounds good. Separated out `JsonApiSchema::Structure` with a `# rubocop:disable Metrics/MethodLength` to silence the last complaint. Still some other tests failing after rebase (before they were green so should be unrelated).

Also, it didnt feel right to leave a double customer query in `#show`, caused by `before_action :authorize_action` invoking the `customer` method. I hotfixed it like below, file diff [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/10330/commits/422be396df270de927af659b7d72001e648a59b2#diff-f77d6cc88c5af9370e8684f7565830dc6050aa3fc22f17a2cc41c8cff1516a9fR60). Squashed into previous commit as it untouches history.

```RUBY
      def customer
        @customer ||= begin 
          if action_name == "show"
            CustomersWithBalance.new(Customer.where(id: params[:id])).query.first || not_found
          else
            Customer.find(params[:id])
          end
        end
      end
```

mkllnk reviewed (approved):
Great!

dacook commented on app/json_schemas/json_api_schema.rb:
```diff
@@ -99,6 +101,59 @@ def data_properties(require_all: false)
+      when Symbol
+        public_send(extra_fields)
+      when Hash
+        extra_fields[:opts] &&
+          public_send(extra_fields[:name], extra_fields[:opts]) || public_send(extra_fields[:name])
```
I wonder if we can use double-splat to simplify this:
```suggestion
        public_send(extra_fields[:name], **extra_fields[:opts])
```

This deconstructs the opts, but I think would have the same end result. It has the benefit of handling blank state so that you don't need to.

mickenorlen replied:
fixed

dacook reviewed (approved):
Great job, thanks so much for working through this one and condensing the git history!

I can see a clear progression and how everything works (it's been a good learning opportunity for me). 
I have a couple of comments but these are just for reference. Looks like a great extension which I'm sure will come in handy for other fields too.

mickenorlen commented on spec/queries/customers_with_balance_spec.rb:
```diff
@@ -18,6 +18,23 @@
+          customers = create_pair(:customer)
+          id_b = described_class.new(Customer.where(id: customers)).query
+            .map{ |c| [c.id, c.balance_value] }
+          expect(id_b).to eq([[customers.first.id, 0], [customers.second.id, 0]])
```
Thank you for looking through this and your comments! Maybe it was a bit quick to approve it since you still had some legit comments here. Appreciate the idea of focusing on the positive though smile

Anyways I added two commits to address your feedback. Since they don't untouch previous history I'm leaving them separated. Let me know if you would prefer squashing them in the end.

mickenorlen replied:
Thank you for looking through this and your comments! Maybe it was a bit quick to approve it since you still had some legit comments here. Appreciate the idea of focusing on the positive though :smile: 

Anyways I added two commits to address your feedback. Since they don't untouch previous history I'm leaving them separated. Let me know if you would prefer squashing them in the end.

dacook reviewed (approved):
💪Great, thanks for doing both those updates, I can see they required a little more thought but the end result is simpler code (yay!). I appreciate the improvements to the comments also  .

Ready for testing!

drummer83 commented:
Hi @mickenorlen,
Thanks for all the work on this one! I have tested this - not being an API expert at all. Here are some notes.

- Running `https://staging.openfoodnetwork.org.uk/api/v1/customers?extra_fields[customer]=balance` as an enterprise admin I get the list of customers for my managed and owned enterprises including the balances. :heavy_check_mark: 
- The balances match the values I can see in the admin area as well. :heavy_check_mark: 
![image](https://user-images.githubusercontent.com/1790176/220336469-76388248-af91-4182-81a1-b7c8d6c657dd.png)
- Running without /extra_fields I get the same list, but the balances are not returned. :heavy_check_mark: 
- I can only see customers of enterprises I own or manage. :x:
![image](https://user-images.githubusercontent.com/1790176/220342404-10f1f158-3efa-4c31-a0c6-2e4ef41cde56.png)
![image](https://user-images.githubusercontent.com/1790176/220342793-388e4809-0764-45c0-9a68-d49873f0fd98.png)
I can also see customers of enterprises which granted permissions to edit their profile, but I am not a manager of. In the example above the user enterprise@ko... (manager of External Producer Shop 0) can see customers of Konrad Producer Shop 2. Is this ok? This was not introduced by this PR, it is present in master already.
- I can create a customer for an enterprise I manage. :heavy_check_mark: 
- I can't create a customer for an enterprise I don't manage. :heavy_check_mark: 
- I can create a customer including `"balance": 1000`. There is no error, but the balance is not written to the database. :question: 
- I can update a customer including `"balance": 1000`. There is no error, but the balance is not written to the database. :question: 
- I can delete a customer for an enterprise I manage. :heavy_check_mark: 
- I can't delete a customer for an enterprise I don't manage. :heavy_check_mark: 
- Reading a single customer only `/api/v1/customers/8289`, the balance is shown WITHOUT using some /extra_fields :question: 

### Questions
@lin-d-hop 
- What about displaying customers of enterprises which I am not owning or a manager of, but I have permission to edit the profile. Is this a bug or intended?
- Is it ok to have no error when creating or updating a customer including a balance, but the balance is not written to the database (see above)?
- Is it ok to always see the balance, when requesting a single customer?

### Conclusion
Adding feedback needed label until feedback from Lynne.
I did not feel very comfortable testing this due to my lack of knowledge. I learned a lot, which was great! To ensure quality we should maybe have someone else have a look at this as well.

mickenorlen commented:
@drummer83 Thank you for testing this. Seems like you've done a really good job in spite of being new to the api. Very clear with the checkboxes, i like it! :D

After checking your findings, it seems to me that everything is still according to specification.

> :x:  I can only see customers of enterprises I own or manage.

Checking the [issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/8909) description this seems according to spec. I should probably have mentioned that there was more info there, sorry about that  :smiley: 

> :question:  I can create a customer including "balance": 1000. There is no error, but the balance is not written to the database.
:question: I can update a customer including "balance": 1000. There is no error, but the balance is not written to the database.

In the issue description they also say no CRUD actions should be possible for balance. In fact it would not be possible as balance is not stored in the customer table but calculated as an aggregate from the customers order history.

As with any unpermitted parameter it is simply filtered out before any action is performed. I would say this is standard behavior.

> :question: Reading a single customer only /api/v1/customers/8289, the balance is shown WITHOUT using some /extra_fields question

Also this is according to specification of both PR and issue description :smile:

mickenorlen commented:
@drummer83 Thank you for testing this. Seems like you've done a really good job in spite of being new to the api. Very clear with the checkboxes, i like it! :D

After checking your findings from a developers perspective, it seems to me that everything is still according to specification. But lets also hear if Lynne will agree.

> :x:  I can only see customers of enterprises I own or manage.

Checking the [issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/8909) description this seems according to spec. I should probably have mentioned that there was more info there, sorry about that  :smiley: 

> :question:  I can create a customer including "balance": 1000. There is no error, but the balance is not written to the database.
:question: I can update a customer including "balance": 1000. There is no error, but the balance is not written to the database.

In the issue description they also say no CRUD actions should be possible for balance. In fact it would not be possible as balance is not stored in the customer table but calculated as an aggregate from the customers order history.

As with any unpermitted parameter it is simply filtered out before any action is performed. I would say this is standard behavior.

> :question: Reading a single customer only /api/v1/customers/8289, the balance is shown WITHOUT using some /extra_fields question

Also this is according to specification of both PR and issue description :smile:

mickenorlen commented:
@drummer83 Thank you for testing this. Seems like you've done a really good job in spite of being new to the api. Very clear with the checkboxes, i like it! :D

After checking your findings from a developers perspective, it seems to me that everything is still according to specification. But lets also hear if Lynne will agree.

> :x:  I can only see customers of enterprises I own or manage.

Checking the [issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/8909) description this seems according to spec. I should probably have mentioned that there was more info there, sorry about that  :smiley: 

> :question:  I can create a customer including "balance": 1000. There is no error, but the balance is not written to the database.
:question: I can update a customer including "balance": 1000. There is no error, but the balance is not written to the database.

In the issue description they also say no CRUD actions should be possible for balance. In fact it would not be possible as balance is not stored in the customer table but calculated as an aggregate from the customers order history.

As with any unpermitted parameter it is simply filtered out before any action is performed. I would say this is standard behavior.

> :question: Reading a single customer only /api/v1/customers/8289, the balance is shown WITHOUT using some /extra_fields

Also this is according to specification of both PR and issue description

mkllnk commented:
> As with any unpermitted parameter it is simply filtered out before any action is performed. I would say this is standard behavior.

This is confusing. The intention of the code is to raise an error and report unprocessable entity. There's a spec for that as well. Now here's a guess of what may be happening:

* Rails can [copy attribute parameters](https://edgeguides.rubyonrails.org/configuring.html#config-action-controller-wrap-parameters-by-default) from the root level to the model level. So `?first_name=Mario` is found at `params[:first_name]` but also copied to `params[:customer][:first_name]`.
* We take take the model attributes, not the global ones: `params.require(:customer).permit(...)` and we don't permit balance here.
* If we tried to access `customer_params[:balance]` it would raise an error, I think.
* But we only use present attributes with `customer.update(customer_params)`.
* And Rails copies only customer attributes and ignores the balance.

In summary, I think, `?customer[balance]=100` would raise an error but `?balance=100` doesn't. We still allow `?first_name=Mario` though. My decision to use parameter wrapping (copying) as in the rest of the app is confusing here.

I'm not sure how we should proceed though. This issue is unrelated to this PR and I think that it's okay to leave the behaviour as is but if someone has a better idea then that would be welcome, too.

dacook commented:
Thanks for investigating Maikel.
I agree that an unpermitted parameter should raise an error. The user needs the feedback that they tried something and it didn't work. But it's probably not that big a deal.
Given that this behaviour isn't added by the PR, I agree we should proceed as-is.

drummer83 commented:
Hi @lin-d-hop,
any chance you can help us moving forward here?
Thanks!

lin-d-hop commented:
Hi folks,
Sorry for the delayed response. I totally missed this :blush: 

_Q1: What about displaying customers of enterprises which I am not owning or a manager of, but I have permission to edit the profile. Is this a bug or intended?_

As a rule of thumb, the functionality of the API should match the functionality of the app. As a user I with permission to edit the profile of a customer, when I go to the customer page I cannt view or edit this customer. I can only view or edit Customer which I own or manage.

Therefore this is a bug. We do not want to be able view or edit customers that we have permission to edit the profile of via the API.

_Q2:Is it ok to have no error when creating or updating a customer including a balance, but the balance is not written to the database (see above)?_

Since customer balance is an aggregated field there can be no updating of customer balance ever by anything a user touches :) But this question is more about error handling. I think on V1 of the API, if I user gives a malformed request we should let them know. My preferred approach would be that if in a Create or Update action a user includes the a value for balance we return an error - _Balance cannot be set or updated._

Having said that, I would be happy to just proceed as is on this and to specify in the API docs that Balanced cannot be created or updated via the API as it is an aggregated field. That would satisfy the requirements for me.

_Q3: Is it ok to always see the balance, when requesting a single customer?_
Yes this was clear in the original spec. We are hiding the balance from the list view because the overheads of balance calculations are pretty severe and can impact performance of the entire app. However this is not a problem for the single customer view and it makes sense to include it by default.

drummer83 commented:
Thanks, @lin-d-hop!

I moved this PR back to In Dev so we can fix Q1. @mickenorlen Will you be up for that?
Q2 can be left as is, just update the API guide.
Q3 is ok.

mickenorlen commented:
Interesting findings, it's amazing to see great minds working together to find the best path forward!

@drummer83 Yes, I'll get back to this next week.

mickenorlen commented:
@drummer83 Sorry I spoke too soon. It turns out I'm not going to be working for OFN so I have to focus my time elsewhere for now. Sorry I couldn't see this one through. A big shoutout to @mkllnk, it really was a pleasure to work with you all. I hope our paths may cross again in the future.

drummer83 commented:
Oh no, @mickenorlen, so sorry to hear...

All the best for you and your projects!

mickenorlen commented:
Thank you so much! There is still a chance I could come back at a later date, but since nothing is certain for now, I really need to consider some other options as well. Contributing to open source is a beautiful thing, but I also have to take care of the family :)

However, after having looked through this a second time. Q1 seems like a separate bug related to API customer selection in general. So I think you could just merge this one in as it is still all green, and open up a new issue. That way we could all have a bit of closure here.

dacook commented:
Hi @RachL I just noticed that you've added the `feedback-needed` label; was this intentional, or is this good to go?

RachL commented:
@dacook yes as we said in delivery circle that we would check if @lin-d-hop was okay with the merge.

sigmundpetersen commented:
What about creating the outstanding issue before we merge, so we don't forget

lin-d-hop commented:
Yes okay, let's merge like this. Looks like an S2 bug on customers since this is a v1 endpoint :-/
I'll add creating the bug to my list.
